src/evaluators/state/PHAL_LoadStateField_Def.hpp

```diff
@@ -50,11 +50,51 @@ void LoadStateFieldBase<EvalT, Traits, ScalarType>::evaluateFields(typename Trai
   //       If dev data has changed, it should be synced to host by
   //       whomever changed the data.
   const auto& stateToLoad = (*workset.stateArrayPtr)[stateName];
-  auto stateData   = stateToLoad.host().data();
-  PHAL::MDFieldIterator<ScalarType> d(data);
-  for (int i = 0; ! d.done() && i < stateToLoad.size(); ++d, ++i)
-    *d = stateData[i];
-  for ( ; ! d.done(); ++d) *d = 0.;
+  Kokkos::deep_copy(stateToLoad.dev(), stateToLoad.host());
+  auto stateData   = stateToLoad.dev();
+  const int stateToLoad_size = stateToLoad.size();
+
+  std::vector<size_t> data_dims;
+  data.dimensions(data_dims);
+
+  const int d1 = data_dims[0];
+  const int d2 = (data_dims.size() > 1) ? data_dims[1] : 0;
+  const int d3 = (data_dims.size() > 2) ? data_dims[2] : 0;
+  const int d4 = (data_dims.size() > 3) ? data_dims[3] : 0;
+  const int d5 = (data_dims.size() > 4) ? data_dims[4] : 0;
+
+  const int data_dims_size = data_dims.size();
+
+  Kokkos::parallel_for(Kokkos::RangePolicy(0,data.size()),
+                       KOKKOS_CLASS_LAMBDA(const int i) {
+    if (data_dims_size == 1) {
+      data(i) = (i < stateToLoad_size) ? stateData(i) : 0.0;
+    } else if (data_dims_size == 2) {
+      int i2 = i % d2;
+      int i1 = i / d2;
+      data(i1,i2) = (i < stateToLoad_size) ? stateData(i) : 0.0;
+    } else if (data_dims_size == 3) {
+      int i3 = i % d3;
+      int i2 = (i / d3) % d2;
+      int i1 = i / (d3 * d2);
+      data(i1,i2,i3) = (i < stateToLoad_size) ? stateData(i) : 0.0;
+    } else if (data_dims_size == 4) {
+      int i4 = i % d4;
+      int i3 = (i / d4) % d3;
+      int i2 = (i / (d4*d3)) % d2;
+      int i1 = i / (d4*d3*d2);
+      data(i1,i2,i3,i4) = (i < stateToLoad_size) ? stateData(i) : 0.0;
+    } else if (data_dims_size == 5) {
+      int i5 = i % d5;
+      int i4 = (i / d5) % d4;
+      int i3 = (i / (d5*d4)) % d3;
+      int i2 = (i / (d5*d4*d3)) % d2;
+      int i1 = i / (d5*d4*d3*d2);
+      data(i1,i2,i3,i4,i5) = (i < stateToLoad_size) ? stateData(i) : 0.0;
+    } else {
+      // Exception
+    }
+  });
```

Comment:
Are you sure the state data indices always unroll the same? Also, this looks like it should be `RangePolicy(0, stateToLoad_size)`? Also you could create 5 lambda kernels here. An exception could also be thrown outside the kernel for `data_dims_size >5`.

Reply:
According to the MDFieldIterator documentation, this should be consistent with how the indices are unrolled. As for the RangePolicy size, I did it this way to be consistent with what we had before. Anything beyond stateToLoad_size will be set to 0.0. This can be changed if we know MDFields get initialized to zero.

I'll go ahead and add that exception.

Reply:
Ah okay I see `MDFieldIterator` is our own thing. I would modify that to work on device so that it's consistent across all code that uses it.

Reply:
It's defined here: https://github.com/sandialabs/Albany/blob/master/src/PHAL_Utilities.hpp#L58 . It needs to be made device friendly or completely removed so that others won't use it.

Reply:
Also, I don't think you should rewrite this code in other evaluators. It should be made generic so that other evaluators can call the same code.